### PR TITLE
Pull decam

### PR DESCRIPTION
**Pull request by rknop**

### CODE REVIEW

File: tests/models/test_image.py
Comment (guynir42) [inline tests/models/test_image.py:107]: Try using `recursive_merge` from `base.py`. I think it may solve these problems (after eating sh*t from this same issue I added something that hopefully solves it).

File: tests/models/test_instrument.py
Comment (guynir42) [inline tests/models/test_instrument.py:303]: Maybe put this in conftest?

File: tests/models/test_instrument.py
Comment (guynir42) [inline tests/models/test_instrument.py:370]: If you decide to change it to `download_and_commit` then rename the test, too.

File: util/radec.py
Comment (guynir42) [inline util/radec.py:1]: This already exists in `pipeline/utils.py`. 
Clearly we have too many util files, and should consider unifying them. 
Perhaps this would be better done in a separate PR. 
For now, you can leave this if you feel these functions are better, and we can decide later which file to merge into the other.

**Review by guynir42 — COMMENTED** [review]
Ahhh only 38 comments. I'm sorry :sweat_smile:

File: models/exposure.py
Comment (rknop) [reply models/exposure.py:308]: So, thinking about this: as things stand right now, we are going to need something like a "create or load" method for Provenance that goes to the database.  Reason: if you just create a Provenance with the constructor, and do things, when you get to committing it to the database, a simple session.merge() will not merge it properly, as the new provenance you created won't have the same primary key as the other provenance that has the unique hash in the database.   That's the whole reason I wrote this; if you ask for a provenance that already exists, you need to get the already existing one.  There's no way to do that without either using a session or creating a session to connect to the database.

There is another solution, which would be my preference.  That would be to take the id field out of Base in models/base.py.  Add a HasPrimaryID mixin that has the auto-incrementing id.  Most models would include that, but Provenance would not; for Provenance, the unique hash would *be* the primary key.  Provenance doesn't need an autoincrementing primary key, since the unique hash already serves any purpose a primary key would need.  Then, we could create provenances by just making objects, and when we merge them with the session it would merge correctly.

File: models/exposure.py
Comment (rknop) [reply models/exposure.py:484]: Yeah... I'll look at it and see if there's something obvious to do.

File: models/exposure.py
Comment (rknop) [reply models/exposure.py:567]: The comment was actually written earlier in my thinking, and I didn't remove it; I'm not sure exactly what the conditions were when I wrote the comment.

nofile serves the "don't download" function of get_fullpath().

File: models/instrument.py
Comment (rknop) [reply models/instrument.py:794]: I only learned recently that you can pass a tuple to isinstance for ORing possible classes.

File: models/instrument.py
Comment (rknop) [reply models/instrument.py:1170]: Experience; a minimum exposure time is something I've needed.  I've used "minexptime" in the past to filter out short "centering" exposures that we didn't really want to process.  (The DECAT-DDF program did this for the first semester, until we decided that the telescope's centering always seemed to be reliable so it wasn't necessary.)  I can't think of a reason why we'd want to filter out long exposures.

File: models/instrument.py
Comment (rknop) [reply models/instrument.py:1189]: Something that would be an instance of collections.abc.Sequence -- lists, tuples, strings.

The real assumption here is that's something that can be indexed with an integer, because some of the other methods take an integer as an argument.

File: models/instrument.py
Comment (rknop) [reply models/instrument.py:1267]: There are a few reasons.  First, I don't want to make any general assumptions about the way each instrument stores its internal list of things found.  Every telescope and archive has its own interface, and this is something we're going to need to customize a lot.  (I have used a class hierarchy for this kind of thing in the past -- intermediate subclasses that capture something that's the same for a number of different instruments.)  While a pandas dataframe might be fine for everything, I'm wary of trying to impose structure on things we don't control, i.e. external interfaces for how to get things from telescopes and archives.  It's always surprisingly heterogeneous.

I also don't want to assume that the way you go from whatever's stored internally to figuring out how you copy the image from the server to the local machine is always going to be the same.  It *might* always be a url... but I wouldn't be surprised if sometimes it turned out it wasn't REST-like, but was a URL where you had to pass some parameters in a POST to get the thing you want.

I think we're best off leaving this as explicitly customizable by instrument, rather than trying to anticipate patterns and put them in a general routine.

File: models/instrument.py
Comment (rknop) [reply models/instrument.py:1271]: Again, only if we can assume that the internal storage is going to be the same for all instruments, which I don't want to assume.

File: models/instrument.py
Comment (rknop) [reply models/instrument.py:1307]: I *suspect* that the chips got named before something changed about how the instrument was going to be mounted to the telescope.  I can't imagine somebody consciously choosing the naming convention they have without either something like that having happened, or without there just being  a mistake.  Who knows.

File: models/instrument.py
Comment (rknop) [reply models/instrument.py:1767]: Deletes the file

File: tests/models/test_instrument.py
Comment (rknop) [reply tests/models/test_instrument.py:303]: My preference is for fixtures that are specific to a given test module to be in that module.  Otherwise, conftest.py gets huge.  conftest.py ought only to have things that are general, or that will be used in multiple different test modules.

File: models/exposure.py
Comment (guynir42) [reply models/exposure.py:237]: Yeah, probably right.

File: models/exposure.py
Comment (guynir42) [reply models/exposure.py:308]: That's an interesting suggestion. I need to think about that.

Comment (guynir42) [reply models/exposure.py:308]: I agree that you need to check the database for Provenance. There's a tool to do that on the data_store object. 
The problem is that this is not a simple thing you can just have happen automatically in the constructor of the parent class. You need to know the upstream provenance and all the parameters.

File: models/instrument.py
Comment (guynir42) [reply models/instrument.py:1170]: As long as we can always add it later, it's fine by me.

File: models/instrument.py
Comment (guynir42) [reply models/instrument.py:1189]: I mean to ask "is the internal storage supposed to be a sequence?" the subject in that sentence isn't clear...

File: models/instrument.py
Comment (guynir42) [reply models/instrument.py:1744]: Then perhaps we should filter the query based on instrument name as well.

File: tests/models/test_instrument.py
Comment (guynir42) [reply tests/models/test_instrument.py:303]: I can imagine using decam exposures in other places.

**Comment by guynir42** [comment]
@rknop let me know when you are done with the previous set of comments (and when you fix all the conflicts) so I will have another review. Thanks!

File: models/instrument.py
Comment (rknop) [reply models/instrument.py:1466]: _chip_radec_off is a variable in the DECam class, not the parent class.

File: tests/models/test_image.py
Comment (guynir42) [inline tests/models/test_image.py:444]: Why do we need these merges now? Does it have to do with the exposure's provenance?

File: tests/models/test_image.py
Comment (guynir42) [inline tests/models/test_image.py:837]:
```suggestion
```

File: util/util.py
Comment (guynir42) [inline util/util.py:31]: I think both "and not" statements could do without parenthesis, but it is just a matter of style so I leave it up to you.

File: util/util.py
Comment (guynir42) [inline util/util.py:3]: I love this very much. I will for sure use this in all my future projects :)

File: models/decam.py
Comment (guynir42) [inline models/decam.py:272]: I can imagine ways where specifying only minmjd could end up a really long query...

File: models/decam.py
Comment (guynir42) [inline models/decam.py:266]: Add missing parameters. Let me just say that using copilot makes this sort of tedium really fun.

**Review by guynir42 — APPROVED** [review]
Looks really good. I can't wait to start using this. 

I have some minor comments, but you can fix them and merge. 

The only big problem I have is the conftest where you delete everything. That seems to me like a hand-grenade waiting to blow up in our face in a few months. I'd rather have leftover test objects on the DB than have complete destruction. I have a few ideas how to do this less dangerously but that would be in another PR.

**Pull request merged**
